[Admin][UI] Sending administrator's password reset email

jakubtobiasz opened the pull request:
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | master <!-- see the comment below -->          |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no <!-- don't forget to update the UPGRADE-*.md file --> |
| Related tickets |                       |
| License         | MIT                                                          |

Continuation of #14128.

What I've managed to done:
- Add `Forgot password?` button on Admin Panel login page
- <img width="611" alt="CleanShot 2022-07-07 at 9 23 57@2x" src="https://user-images.githubusercontent.com/80641364/177855856-a03376c2-77e4-44c8-b8f0-bab324b3256f.png">
- Add request admin's password reset page
- <img width="687" alt="CleanShot 2022-07-07 at 9 24 08@2x" src="https://user-images.githubusercontent.com/80641364/177855877-45c4b510-0100-4eae-a359-56e498b413ed.png">
- Add tests checking if the correct version of email is send for both API and UI part

And.. flash message 💃🏻.
<img width="593" alt="CleanShot 2022-07-07 at 9 25 20@2x" src="https://user-images.githubusercontent.com/80641364/177856062-7b0ea4ab-be35-4281-ba0d-0e02b23f78aa.png">

jakubtobiasz committed:
Define behaviors for Sending an administrator's password reset request from UI

jakubtobiasz committed:
Add RequestPasswordReset page

jakubtobiasz committed:
Add handling requested password request sent from UI

jakubtobiasz committed:
Add tests checking if the correct email is sent for both Admin and Api bundles

jakubtobiasz committed:
Add "Forgot password?" button on the login page to Admin Panel

jakubtobiasz committed:
Change the text searched for while asserting if email has been sent

jakubtobiasz committed:
Add request password reset button on Admin Panel login page

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Action/Account/PostRequestPasswordResetAction.php:
```diff
@@ -0,0 +1,56 @@
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+final class PostRequestPasswordResetAction
```
```suggestion
final class AdminResetPasswordAction
```
I feel suffix with the type of operation is redundant, but WDYT?

Zales0123 replied:
You mean... `*Action`? 😄 this is the convention we have in all one-action controllers

jakubtobiasz replied:
@Zales0123 I think he meant `Post` part but I think it's not redundant as the same route has separate action for `Get` and `Post` action so it was easier to put method name than inventing a new name less related to the thing it's about.

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Resources/config/routing/security.yml:
```diff
@@ -17,3 +17,13 @@ sylius_admin_login_check:
     methods: [GET]
+
+sylius_admin_request_password_reset:
+    path: /forgotten-password
```
```suggestion
    path: /reset-password
```
why no simple reset password?

jakubtobiasz replied:
@AdamKasp The same route was on shop side, I wanted stick to convention.

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Resources/config/services/controller.xml:
```diff
@@ -39,6 +39,20 @@
+            <argument type="service" id="session.flash_bag"/>
+            <argument type="service" id="router"/>
+
+            <tag name="controller.service_arguments" />
```
why do we need this tag, when service is already public?

jakubtobiasz replied:
@AdamKasp As I remember it passes `Request` obj as an argument to `__invoke()` method. I'll check if I'm right, if it turns out it's redundant I'll remove it.

jakubtobiasz replied:
I was wrong, I'll remove it :D.

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Resources/config/services/controller.xml:
```diff
@@ -39,6 +39,20 @@
             <argument type="service" id="session" />
         </service>
 
+        <service id="Sylius\Bundle\AdminBundle\Action\Account\GetRequestPasswordResetAction" public="true">
```
```suggestion
        <service id="Sylius\Bundle\AdminBundle\Action\Account\GetRequestPasswordResetAction">
```
all services in this file are public by default (line 20)

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Form/Model/PasswordResetRequest.php:
```diff
@@ -0,0 +1,34 @@
+        return $this->email;
+    }
+
+    public function setEmail(?string $email): void
```
do we need this setter? this class looks like VO, can we set this email in constructor?

jakubtobiasz replied:
@AdamKasp Before executing `$form->handleRequest($request);` the object is "empty" so forcing to set e-mail as a constructor's parameter in my opinion is not the best solution

Zales0123 commented on src/Sylius/Bundle/AdminBundle/Tests/MessageHandler/Admin/SendResetPasswordEmailHandlerTest.php:
```diff
@@ -0,0 +1,74 @@
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class SendResetPasswordEmailHandlerTest extends KernelTestCase
```
I'm wondering do we need this test 🤔 It's well covered with Behat scenarios, as far as I see. It also has spec already. So we can safely skip it IMO 🖖 (even though generally more tests -> better 🔥)

jakubtobiasz replied:
Behat is testing the application as a whole (with ApiBundle, ShopBundle and AdminBundle). I had a talk with Łukasz we aren't able to check if system is sending version A of reset password email (in case AdminBundle is loaded) or version B (when there's only ApiBundle). He advised me to write integration tests directly inside `AdminBundle` and `ApiBundle`. If you check the last assert you'll notice we're checking against different strings in an email's content.

Zales0123 commented on src/Sylius/Bundle/AdminBundle/composer.json:
```diff
@@ -35,7 +35,9 @@
         "twig/intl-extra": "^2.12",
-        "twig/twig": "^2.12"
+        "twig/twig": "^2.12",
+        "knplabs/knp-menu": "^3.1",
+        "knplabs/knp-menu-bundle": "^3.0"
```
Alphabetical order R.I.P [*]

Zales0123 replied:
Btw, why do we need it added?

jakubtobiasz replied:
@Zales0123 To run integration tests (why they were needed you can read in comment related to [src/Sylius/Bundle/AdminBundle/Tests/MessageHandler/Admin/SendResetPasswordEmailHandlerTest.php](https://github.com/Sylius/Sylius/pull/14138/files/6cd26878ccd8887421cf94617390a896d2df781f#diff-ee0687c388c61d7aa9653ad8091b0840adefdb1653a44fdc10e6627b30fa6008). I needed `Kernel` with loaded `AdminBundle` without `ApiBundle`. `AdminBundle` was complaining about missing service from `knp-menu-bundle` so I've added it.

Zales0123 commented on src/Sylius/Bundle/AdminBundle/spec/Action/Account/GetRequestPasswordResetActionSpec.php:
```diff
@@ -0,0 +1,54 @@
+use Symfony\Component\HttpFoundation\Response;
+use Twig\Environment;
+
+final class GetRequestPasswordResetActionSpec extends ObjectBehavior
```
We do not need this spec for sure :dancer: controllers (or controllers actions) should be as simple as possible and all the functionality should be covered by Behat scenarios. _Specing_ controller can lead to justifying putting there too much logic, which we definitely want to avoid 🖖

Zales0123 commented on src/Sylius/Bundle/AdminBundle/spec/Action/Account/PostRequestPasswordResetActionSpec.php:
```diff
@@ -0,0 +1,78 @@
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+final class PostRequestPasswordResetActionSpec extends ObjectBehavior
```
Same here

Zales0123 commented on src/Sylius/Bundle/ApiBundle/Tests/MessageHandler/Admin/SendResetPasswordEmailHandlerTest.php:
```diff
@@ -0,0 +1,74 @@
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class SendResetPasswordEmailHandlerTest extends KernelTestCase
```
Same as above 🖖

jakubtobiasz commented on src/Sylius/Bundle/AdminBundle/Action/Account/GetRequestPasswordResetAction.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function __invoke(): Response
+    {
+        $form = $this->formFactory->create(RequestPasswordResetType::class);
```
Why? If anybody need to change anything in this FormType there are form extensions 🤔.

Zales0123 replied:
Ok, it makes sense 👍

jakubtobiasz commented on src/Sylius/Bundle/AdminBundle/Action/Account/PostRequestPasswordResetAction.php:
```diff
@@ -0,0 +1,56 @@
+
+        $this->flashBag->set('success', 'sylius.admin.request_reset_password.success');
+
+        return new RedirectResponse($this->router->generate('sylius_admin_login'));
```
It can be decorated 🤔. I have mixed feeling about having too many different parameters.

jakubtobiasz commented on src/Sylius/Bundle/AdminBundle/Action/Account/PostRequestPasswordResetAction.php:
```diff
@@ -0,0 +1,56 @@
+                $formData->getEmail()
+            );
+
+            $this->messageBus->dispatch($requestPasswordResetMessage);
```
What kind of problems do you mean?

jakubtobiasz replied:
I'll handle it with validation in upcoming PR :).

jakubtobiasz commented on src/Sylius/Bundle/AdminBundle/Action/Account/PostRequestPasswordResetAction.php:
```diff
@@ -0,0 +1,56 @@
+            $this->messageBus->dispatch($requestPasswordResetMessage);
+        }
+
+        $this->flashBag->set('success', 'sylius.admin.request_reset_password.success');
```
I'll handle it with validation in upcoming PR :).

jakubtobiasz committed:
Remove redundant making services public and reorder composer packages

jakubtobiasz committed:
Rename password reset related actions

jakubtobiasz committed:
Allow to define redirect route from routes declaration

jakubtobiasz committed:
Remove redundant action test

jakubtobiasz commented:
@Zales0123 
All comments resolved. I've added an option to define redirect route from yml. I've decided to put "the default" value in a code anyway just to be sure we'll never spot `null` there. Also I've removed spec for `Render` action but after implementing the `define redirect route` option I've decided to left the second action. If you still feel we should remove it just let me know 🖖🏼.

@AdamKasp 
Most of your comments are resolved, just 2 need your interaction :D.

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Action/Account/RenderResetPasswordPageAction.php:
```diff
@@ -0,0 +1,39 @@
+use Symfony\Component\HttpFoundation\Response;
+use Twig\Environment;
+
+final class RenderResetPasswordPageAction
```
IMO `RenderRequestPasswordResetPageAction` would better convey what page this controller actually refers to.

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Action/Account/RequestPasswordResetAction.php:
```diff
@@ -0,0 +1,57 @@
+        $this->flashBag->set('success', 'sylius.admin.request_reset_password.success');
+
+        $options = $request->attributes->get('_sylius');
+        return new RedirectResponse($this->router->generate($options['redirect'] ?? 'sylius_admin_login'));
```
IIRC the `redirect` attribute can also be an array.

GSadee replied:
```suggestion

        return new RedirectResponse($this->router->generate($options['redirect'] ?? 'sylius_admin_login'));
```

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Form/Model/PasswordResetRequest.php:
```diff
@@ -0,0 +1,34 @@
+
+    public function __construct()
+    {
+        $this->email = null;
```
What's the point of this constructor when you can just initialize it inline?

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Form/RequestPasswordResetType.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $resolver->setDefaults([
+            'data_class' => PasswordResetRequest::class,
+            'csrf_token_id' => 'reset_password',
```
Is setting a specific token id actually necessary?

GSadee commented on src/Sylius/Bundle/AdminBundle/Action/Account/RequestPasswordResetAction.php:
```diff
@@ -0,0 +1,57 @@
+
+    public function __invoke(Request $request): RedirectResponse
+    {
+        $requestPasswordReset = new PasswordResetRequest();
+        $form = $this->formFactory->create(RequestPasswordResetType::class, $requestPasswordReset);
```
Is it needed here?
```suggestion
        $form = $this->formFactory->create(RequestPasswordResetType::class);
```

GSadee commented on src/Sylius/Bundle/AdminBundle/spec/Action/Account/RequestPasswordResetActionSpec.php:
```diff
@@ -0,0 +1,115 @@
+        $response->getTargetUrl()->shouldReturn('/login');
+    }
+
+    public function it_should_redirect_to_default_route_if_custom_one_is_not_defined(
```
```suggestion
    public function it_redirects_to_default_route_if_custom_one_is_not_defined(
```

jakubtobiasz committed:
Add handling sylius redirect parameter's array form

jakubtobiasz committed:
Rename RenderResetPasswordPageAction to RenderRequestPasswordResetPageAction

jakubtobiasz commented:
@GSadee @NoResponseMate thanks for the suggestions, all introduced 👊🏻.

GSadee commented on src/Sylius/Bundle/AdminBundle/Form/RequestPasswordResetType.php:
```diff
@@ -0,0 +1,45 @@
+
+    public function getBlockPrefix(): string
+    {
+        return 'sylius_admin_render_reset_password_page';
```
```suggestion
        return 'sylius_admin_request_password_reset';
```

GSadee commented on src/Sylius/Bundle/AdminBundle/Resources/config/app/events.yaml:
```diff
@@ -1287,3 +1287,27 @@ sylius_ui:
+            blocks:
+                form:
+                    template: '@SyliusAdmin/Security/RequestPasswordReset/_form.html.twig'
+                    priority: 30
+
+        sylius.admin.request_password_reset.before_form:
+            blocks:
+                header:
+                    template: "@SyliusAdmin/Security/_logo.html.twig"
+                    priority: 30
+                flashes:
+                    template: "@SyliusUi/_flashes.html.twig"
+                    priority: 20
+
+        sylius.admin.request_password_reset.form.content:
+            blocks:
+                email:
+                    template: '@SyliusAdmin/Security/RequestPasswordReset/_email.html.twig'
+                    priority: 30
+                submit:
+                    template: '@SyliusAdmin/Security/RequestPasswordReset/_submit.html.twig'
+                    priority: 20
```
```suggestion
                    priority: 30

        sylius.admin.request_password_reset.before_form:
            blocks:
                header:
                    template: "@SyliusAdmin/Security/_logo.html.twig"
                    priority: 20
                flashes:
                    template: "@SyliusUi/_flashes.html.twig"
                    priority: 10

        sylius.admin.request_password_reset.form.content:
            blocks:
                email:
                    template: '@SyliusAdmin/Security/RequestPasswordReset/_email.html.twig'
                    priority: 10
                submit:
                    template: '@SyliusAdmin/Security/RequestPasswordReset/_submit.html.twig'
                    priority: 10
```

GSadee commented on src/Sylius/Bundle/AdminBundle/Resources/config/validation/PasswordResetRequest.xml:
```diff
@@ -0,0 +1,21 @@
+-->
+
+<constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
+    <class name="Sylius\Bundle\AdminBundle\Form\Model\PasswordResetRequest">
+        <property name="email">
+            <constraint name="NotBlank" />
+            <constraint name="Email" />
+        </property>
+    </class>
```
Is there an implemented scenario for this validation?

jakubtobiasz replied:
Nope, I've split the main task a little bit too much and validation will be covered later.

GSadee commented on config/packages/security.yaml:
```diff
@@ -110,6 +110,8 @@ security:
         - { path: "%sylius.security.shop_regex%/_partial", role: IS_AUTHENTICATED_ANONYMOUSLY, ips: [127.0.0.1, ::1] }
         - { path: "%sylius.security.shop_regex%/_partial", role: ROLE_NO_ACCESS }
 
+        - { path: "%sylius.security.admin_regex%/forgotten-password", role: IS_AUTHENTICATED_ANONYMOUSLY }
```
This change should be added to the UPGRADE file

GSadee commented on src/Sylius/Behat/Context/Api/EmailContext.php:
```diff
@@ -156,11 +156,19 @@ public function aWelcomingEmailShouldHaveBeenSentTo(string $recipient, string $l
     }
 
+    /**
+     * @Then /^"([^"]+)" should receive no emails$/
```
```suggestion
     * @Then :recipient should receive no emails
```

GSadee reviewed (commented):
If I see correctly, you've added validation for both API and UI to this PR, IMO it would be better to add another PR with validation as you've planned before

jakubtobiasz committed:
Rename RequestPasswordResetType block prefix

jakubtobiasz committed:
Update block priorities in template events

jakubtobiasz committed:
Add information about required changes in security.yaml to UPGRADE files

jakubtobiasz commented:
@GSadee, after a conversation with @NoResponseMate I've dropped his three commits and I've resolved all your comments. While adding information about changes in `security.yaml` I noticed we didn't mention a similar change in the case of API. I've included it in the same PR as all API-related PR are already merged ✌🏻.

GSadee commented on UPGRADE-1.12.md:
````diff
@@ -19,6 +19,15 @@ and should be used only this way.
       )
     ```
 
+4. To allow administrator reset his password, add in `config/packages/security.yaml` file the following entry
````
```suggestion
4. To allow administrator reset their password, add in `config/packages/security.yaml` file the following entry
```

GSadee commented on UPGRADE-API-1.12.md:
```diff
@@ -100,3 +100,12 @@ Here is how the response looks like:
 
 1. The responses of endpoints `/api/v2/admin/products` and `/api/v2/admin/products/{code}` have been changed in such a way that the field `defaultVariant` has been removed.
+
+1. To allow administrator reset his password, add in `config/packages/security.yaml` file the following entry
```
I guess it has been already added in this file: https://github.com/Sylius/Sylius/pull/14149

GSadee commented:
Thank you, Jakub! :1st_place_medal:
